sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1480,10 +1522,50 @@ _pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
   if (ZeResult) {
     die("[L0] getZeQueue: failed to create queue");
   }
+
+  // When using immediate commandlists, we create a commandlist corresponding
+  // to the L0 queue created above. Created once, the commandlists are reused
+  // without needing a close or reset.
+  if (UseImmediateCommandLists) {
+    ze_command_list_handle_t ZeCommandList;
+    ZE_CALL_NOCHECK(zeCommandListCreateImmediate,
+                    (Queue->Context->ZeContext, Queue->Device->ZeDevice,
+                     &ZeCommandQueueDesc, &ZeCommandList));
+    ImmCmdLists[CurrentIndex] =
+        Queue->CommandListMap
+            .insert(std::pair<ze_command_list_handle_t, pi_command_list_info_t>{
+                ZeCommandList,
+                {nullptr, true, ZeQueues[CurrentIndex], *QueueGroupOrdinal}})
+            .first;
+    // Add this commandlist to the cache so it can be destroyed as part of
+    // QueueRelease
+    auto &ZeCommandListCache =
+        QueueType != queue_type::Compute
+            ? Queue->Context->ZeComputeCommandListCache[Queue->Device->ZeDevice]
+            : Queue->Context->ZeCopyCommandListCache[Queue->Device->ZeDevice];
+    std::lock_guard<std::mutex> lock(Queue->Context->ZeCommandListCacheMutex);
+    ZeCommandListCache.push_back(ZeCommandList);
+  }
+
   return ZeQueue;
 }
 
+// This function will return one of possibly multiple available
+// immediate commandlists associated with this Context/Device.
```

Comment:
I thought we were going to use single immediate command-list (index 0) for all submissions, why multiple?

Reply:
We can do either.
The L0 team expects a slight benefit from using all compute queues as well as all copy queues.

Reply:
Let's go with a single one first, and only add overhead of using multiple ones when we understand the exact scenarios where it is better.

Reply:
What about this comment? We don't want the overhead of multiple immediate command-lists unless there is a compelling case where it is necessary (please leave a TODO comment).